numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -1385,21 +1385,23 @@ array_datetime_as_string(PyObject *NPY_UNUSED(self), PyObject *args,
     /* Parse the input unit if provided */
     if (unit_in != NULL && unit_in != Py_None) {
         PyObject *strobj;
-        char *str = NULL;
-        Py_ssize_t len = 0;
 
-        if (PyUnicode_Check(unit_in)) {
-            strobj = PyUnicode_AsASCIIString(unit_in);
-            if (strobj == NULL) {
-                goto fail;
+        if (PyBytes_Check(unit_in)) {
+            /* accept bytes input */
+            PyObject *obj_str = PyUnicode_FromEncodedObject(unit_in, NULL, NULL);
+            if (obj_str == NULL) {
+                return 0;
             }
+            strobj = obj_str;
         }
         else {
+            Py_INCREF(unit_in);
             strobj = unit_in;
-            Py_INCREF(strobj);
         }
 
-        if (PyBytes_AsStringAndSize(strobj, &str, &len) < 0) {
+        Py_ssize_t len = 0;
```

Comment:
```suggestion
        Py_ssize_t len;
```
Doesn't matter though.